scubainit/src/main.rs

```diff
@@ -0,0 +1,347 @@
+                return Ok(());
+            }
+        }
+
+        // Okay, add shadow entry
+        let entry = shadow::ShadowEntry {
+            name: user_name.to_owned(),
+            passwd: INVALID_PASSWORD.to_owned(),
+            last_change_date: None,
+            min_password_age: None,
+            max_password_age: None,
+            warn_period: None,
+            inact_period: None,
+            expire_date: None,
+        };
+        let mut writer = shadow::ShadowFileWriter::new(&file);
+        Ok(writer.write(&entry)?)
+    }
+
+    pub fn change_user(&self) -> Result<()> {
+        let uid = self.uid;
+        let gid = self.gid;
+        let user = &self.user;
+        debug!("Changing user to uid={uid}, gid={gid}");
+        unsafe {
+            libc_result(libc::setgroups(0, std::ptr::null()))?;
+            libc_result(libc::setgid(gid))?;
+            libc_result(libc::setuid(uid))?;
+        }
+
+        let home_dir_path = self.home_dir();
+        let home_dir_str = home_dir_path.to_str().context("Invalid home_dir")?;
+        env::set_var("USER", user);
+        env::set_var("LOGNAME", user);
+        env::set_var("HOME", &home_dir_str);
+        Ok(())
+    }
+}
+
+struct Context {
+    user_info: Option<UserInfo>,
+    umask: Option<u32>,
+    user_hook: Option<String>,
+    root_hook: Option<String>,
+}
+
+impl Context {
+    pub fn set_umask(&self) {
+        if let Some(umask) = self.umask {
+            //  SAFETY: This function is not actually unsafe.
+            unsafe {
+                libc::umask(umask);
+            }
+        }
+    }
+
+    pub fn call_root_hook(&self) -> Result<()> {
+        self.call_hook(&self.root_hook)
+    }
+
+    pub fn call_user_hook(&self) -> Result<()> {
+        self.call_hook(&self.user_hook)
+    }
+
+    fn call_hook(&self, path_str: &Option<String>) -> Result<()> {
+        let path = match path_str {
+            None => return Ok(()),
+            Some(x) => x,
+        };
+
+        make_executable(path)?;
+
+        debug!("Executing hook {path}");
+        let status = Command::new(path).status()?;
+        if !status.success() {
+            if let Some(code) = status.code() {
+                bail!("{path} exited with status {code}")
+            }
+            if let Some(signal) = status.signal() {
+                bail!("{path} terminated by signal {signal}")
+            }
+            bail!("{path} exited for an unknown reason! ({status})")
+        }
+        Ok(())
+    }
+}
+
+fn process_envvars_user_info() -> Result<Option<UserInfo>> {
+    // The following variables are optional, but if any is set, all must be set:
+    let uid = pop_env_uint(SCUBAINIT_UID)?;
+    let gid = pop_env_uint(SCUBAINIT_GID)?;
+    let user = pop_env_str(SCUBAINIT_USER);
+    let group = pop_env_str(SCUBAINIT_GROUP);
+
+    match count([
+        uid.is_some(),
+        gid.is_some(),
+        user.is_some(),
+        group.is_some(),
+    ]) {
```

Comment:
This code has evolved a lot since I first wrote it:
```rust
let num_set = (
    uid.is_some() as u32,
    gid.is_some() as u32,
    user.is_some() as u32,
    group.is_some() as u32,
);
if num_set == 4 {
```

I didn't like the repetition of the `as u32` and the hard-coded `4` (what if vars are added/removed from this required set?).

I agree the separate type+function is overkill. I implemented a variation of your suggestion, binding the vector of `is_some()`s, so I could compare against its `len()`.